optimum/fx/optimization/transformations.py

```diff
@@ -0,0 +1,308 @@
+    """
+    Transformation that changes truediv nodes to multiplication by the inverse when the denominator is static.
+    For example, that is sometimes the case for the scaling factor in attention layers.
+    """
+
+    def transform(self, graph_module: "GraphModule", lint_and_recompile: bool = True) -> "GraphModule":
+        graph = graph_module.graph
+        for node in graph.nodes:
+            if node.op == "call_function" and node.target == operator.truediv:
+                x, y = node.args
+                if not isinstance(y, torch.fx.Node):
+                    node.target = operator.mul
+                    node.args = (x, 1 / y)
+                    node.was_transformed = "ChangeTrueDivToMulByInverse"
+
+        if lint_and_recompile:
+            graph.lint()
+            graph_module.recompile()
+        return graph_module
+
+    def reverse(self, graph_module: "GraphModule", lint_and_recompile: bool = True) -> "GraphModule":
+        graph = graph_module.graph
+        for node in graph.nodes:
+            if getattr(node, "was_transformed", "") == "ChangeTrueDivToMulByInverse":
+                node.target = operator.truediv
+                x, y = node.args
+                node.args = (x, 1 / y)
+
+        if lint_and_recompile:
+            graph.lint()
+            graph_module.recompile()
+        return graph_module
+
+
+class DeepCopy(ReversibleTransformation):
+    """
+    Transformation that does nothing except making a deepcopy of the graph module.
+    """
+
+    def transform(self, graph_module: "GraphModule", lint_and_recompile: bool = True) -> "GraphModule":
+        clone = copy.deepcopy(graph_module)
+        # This is needed because copy.deepcopy does not take care of it.
+        # Without these attributes, the reverse transformation cannot be done.
+        for n1, n2 in zip(graph_module.graph.nodes, clone.graph.nodes):
+            if hasattr(n1, "was_transformed"):
+                n2.was_transformed = n1.was_transformed
+        return clone
+
+    def reverse(self, graph_module: "GraphModule", lint_and_recompile: bool = True) -> "GraphModule":
+        return self.transform(graph_module, lint_and_recompile=lint_and_recompile)
+
+
+class LintAndRecompile(ReversibleTransformation):
+    """
+    Transformation that does nothing except linting and recompiling the graph module.
+    """
+
+    def transform(self, graph_module: "GraphModule", lint_and_recompile: bool = True) -> "GraphModule":
+        graph_module.graph.lint()
+        graph_module.recompile()
+        return graph_module
+
+    def reverse(self, graph_module: "GraphModule", lint_and_recompile: bool = True) -> "GraphModule":
+        graph_module.graph.lint()
+        graph_module.recompile()
+        return graph_module
+
+
+def compose(*args: Transformation, inplace: bool = True) -> Callable[["GraphModule"], "GraphModule"]:
+    """
+    Composes a list of transformations together.
+    """
+    transformations = list(reversed(args))
```

Comment:
Why do we need to reverse the list of transformations here?

Reply:
That is because doing `composition = compose(f, g, h)` without reversing the list, the composition corresponds to `f o g o h`. So by reversing, the composition order happens in the same order as the one provided in the arguments.